### PR TITLE
Issue 757: Vectorise GP stan code

### LINKED ISSUE #757: Vectorise GPs
See #742

### PR DESCRIPTION
**Pull request by seabbs**
## Description

This PR builds on #741 and closes #757 by vectorising the spectral density functions, using a single matern function, and introducing periodic kernals. It uses ideas found in https://avehtari.github.io/casestudies/Motorcycle/motorcycle_gpcourse.html.

It takes advantage of this refactor to do some prior predictive checking for the priors related to the GP, and then leverages the improved performance to reduce stan settings and relax initialisation constraints (see the news section). 

## Initial submission checklist

- [x] My PR is based on a package issue and I have explicitly linked it.
- [x] I have tested my changes locally (using `devtools::test()` and `devtools::check()`).
- [x] I have added or updated unit tests where necessary.
- [x] I have updated the documentation if required and rebuilt docs if yes (using `devtools::document()`).
- [x] I have followed the established coding standards (and checked using `lintr::lint_package()`).
- [x] I have added a news item linked to this PR.

## After the initial Pull Request 

- [x] I have reviewed Checks for this PR and addressed any issues as far as I am able.

### CODE REVIEW

File: epinow-epinow-1.png
Comment (seabbs) [inline epinow-epinow-1.png:1]: I removed these but now I see they are main? Surely this can't be right

Comment (seabbs) [reply epinow-epinow-1.png:1]: I've sorted this out in #743

**Comment by seabbs** [comment]
@jamesmbaazam appreciate there is quite a lot going on here. I recommend running through a few models locally to get a sense of the impact of changes. Ideally the prior changes should have minimal impact in most cases and if they do pull behaviour towards being more reasonable. 

The GP change should be equivalent for the same kernals. we might want to mark periodic kernels as experimental as on their own I am not sure of the use case (I plan to add kernel composition soon which will make these make more sense).

**Comment by jamesmbaazam** [comment]
Sorry I didn't get back to you sooner. I will review this early next week. I was working on the `doxygen` stuff on epinowcast, which took much longer than expected, especially with the set up on mac.

File: R/create.R [hunk 1]
```diff
@@ -372,33 +373,50 @@ create_gp_data <- function(gp = gp_opts(), data) {
...
+  if (data$stationary == 1) {
+    time <- time - 1
   }
 
+  obs_time <- data$t - data$seeding_time
+  if (gp$ls_max > obs_time) {
+    gp$ls_max <- obs_time
+  }
+
+  times <- 1:time
```
Comment (jamesmbaazam) [inline R/create.R:390]:
```suggestion
  times <- seq_len(time)
```

File: inst/stan/functions/delays.stan [hunk 2]
```diff
@@ -89,15 +89,10 @@ void delays_lp(vector delay_params,
...
-        ) * weight[d];
-        // if a distribution with postive support only truncate the prior
-        if (delay_dist[d] == 1) {
-          target += -normal_lccdf(
-            0 | delay_params_mean[s], delay_params_sd[s]
-          ) * weight[d];
+        if (weight[d] > 1) {
+          target += weight[d] * normal_lpdf(delay_params[s] | delay_params_mean[s], delay_params_sd[s]);
+        }else {
+          delay_params[s] ~ normal(delay_params_mean[s], delay_params_sd[s]);
```
Comment (jamesmbaazam) [inline inst/stan/functions/delays.stan:95]: Nice!

File: tests/testthat/test-create_gp_data.R [hunk 3]
```diff
@@ -0,0 +1,51 @@
+test_that("create_gp_data returns correct default values when GP is disabled", {
+  data <- list(t = 30, seeding_time = 7, horizon = 7, future_fixed = 0, fixed_from = 0)
+  restricted_time <- 30 - 7 - 1
+  times <- 1:restricted_time
```
Comment (jamesmbaazam) [inline tests/testthat/test-create_gp_data.R:4]:
```suggestion
  times <- seq_len(restricted_time)
```

File: R/opts.R [hunk 4]
```diff
@@ -401,39 +401,43 @@ backcalc_opts <- function(prior = c("reports", "none", "infections"),
...
 #' @param kernel Character string, the type of kernel required. Currently
-#' supporting the squared exponential kernel ("se", or "matern" with
-#' 'matern_order = Inf'), 3 over 2 oder 5 over 2 Matern kernel ("matern", with
-#' `matern_order = 3/2` (default) or `matern_order = 5/2`, respectively), or
-#' Orstein-Uhlenbeck ("ou", or "matern" with 'matern_order = 1/2'). Defaulting
-#' to the Matérn 3 over 2 kernel for a balance of smoothness and
-#' discontinuities.
+#' supporting the squared exponential kernel ("se"), periodic kernel
+#' ("periodic"), Ornstein-Uhlenbeck kernel ("ou"), and the
+#' Matern kernel ("matern").
```
Comment (jamesmbaazam) [inline R/opts.R:432]: Listing them in the order they appear as options in the function.

```suggestion
#' supporting the Matern kernel ("matern"), squared exponential kernel ("se"), periodic kernel, Ornstein-Uhlenbeck #' kernel ("ou"), and the periodic kernel ("periodic").
#' 
```

File: R/opts.R [hunk 5]
```diff
@@ -455,21 +462,30 @@ backcalc_opts <- function(prior = c("reports", "none", "infections"),
...
                     ls_mean = 21,
                     ls_sd = 7,
                     ls_min = 0,
                     ls_max = 60,
-                    alpha_sd = 0.05,
-                    kernel = c("matern", "se", "ou"),
+                    alpha_mean = 0,
+                    alpha_sd = 0.01,
+                    kernel = c("matern", "se", "ou", "periodic"),
                     matern_order = 3 / 2,
```
Comment (jamesmbaazam) [inline R/opts.R:477]: Making the options explicit.
```suggestion
                    matern_order = c(3 / 2, 1 / 2, 5 /2),
```

**Review by jamesmbaazam — CHANGES_REQUESTED** [review]
Thanks @seabbs. I've now reviewed the code aspects as much as I can. I am currently exploring the various gp kernels locally and will report if I find anything worth discussing.

File: R/estimate_infections.R [hunk 6]
```diff
@@ -101,8 +101,7 @@
 #' def <- estimate_infections(reported_cases,
 #'   generation_time = gt_opts(generation_time),
 #'   delays = delay_opts(incubation_period + reporting_delay),
-#'   rt = rt_opts(prior = list(mean = 2, sd = 0.1)),
-#'   stan = stan_opts(control = list(adapt_delta = 0.95))
```
[the remark this answers is not in the document]
Comment (seabbs) [reply R/estimate_infections.R:105]: because we don't want to set anything other than the default options here now

File: R/opts.R (hunk 5 above)
Comment (seabbs) [reply R/opts.R:477]: As its numeric I don't think that will work. It also isn't limited to those options

**Review by jamesmbaazam — APPROVED** [review]

**Pull request merged**